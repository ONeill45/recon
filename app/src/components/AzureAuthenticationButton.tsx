import React, { useState } from 'react'
import {
  Configuration,
  LogLevel,
  PublicClientApplication,
  AuthenticationResult,
  AccountInfo,
  EndSessionRequest,
  RedirectRequest,
  PopupRequest,
} from '@azure/msal-browser'

const AzureActiveDirectoryAppClientId = '8570b3ce-8294-4dba-9048-3f1f9f2eb7ad'

const MSAL_CONFIG: Configuration = {
  auth: {
    authority:
      'https://login.microsoftonline.com/7f7697bc-3ee2-48f2-9d35-7cb75bddd74b',
    clientId: AzureActiveDirectoryAppClientId,
  },
  cache: {
    cacheLocation: 'sessionStorage',
    storeAuthStateInCookie: false,
  },
  system: {
    loggerOptions: {
      loggerCallback: (level, message, containsPii) => {
        if (containsPii) {
          return
        }
        switch (level) {
          case LogLevel.Error:
            console.error(message)
            return
          case LogLevel.Info:
            console.info(message)
            return
          case LogLevel.Verbose:
            console.debug(message)
            return
          case LogLevel.Warning:
            console.warn(message)
            return
        }
      },
    },
  },
}

class AzureAuthenticationContext {
  private myMSALObj: PublicClientApplication = new PublicClientApplication(
    MSAL_CONFIG,
  )
  private account?: AccountInfo
  private loginRedirectRequest?: RedirectRequest
  private loginRequest?: PopupRequest

  public isAuthenticationConfigured = false

  constructor() {
    // @ts-ignore
    this.account = null
    this.setRequestObjects()
    if (MSAL_CONFIG?.auth?.clientId) {
      this.isAuthenticationConfigured = true
    }
  }

  private setRequestObjects(): void {
    this.loginRequest = {
      scopes: [],
      prompt: 'select_account',
    }

    this.loginRedirectRequest = {
      ...this.loginRequest,
      redirectStartPage: 'http://localhost:3000', //window.location.href,
    }
  }

  login(signInType: string, setUser: any): void {
    if (signInType === 'loginPopup') {
      this.myMSALObj
        .loginPopup(this.loginRequest)
        .then((resp: AuthenticationResult) => {
          this.handleResponse(resp, setUser)
        })
        .catch((err) => {
          console.error(err)
        })
    } else if (signInType === 'loginRedirect') {
      this.myMSALObj.loginRedirect(this.loginRedirectRequest)
    }
  }

  logout(account: AccountInfo): void {
    const logOutRequest: EndSessionRequest = {
      account,
    }

    this.myMSALObj.logout(logOutRequest)
  }
  handleResponse(response: AuthenticationResult, incomingFunction: any) {
    if (response !== null) {
      // @ts-ignore
      this.account = response.account
    } else {
      this.account = this.getAccount()
    }

    if (this.account) {
      incomingFunction(this.account)
    }
  }
  // @ts-ignore
  private getAccount(): AccountInfo | undefined {
    console.log('loadAuthModule')
    const currentAccounts = this.myMSALObj.getAllAccounts()
    if (currentAccounts === null) {
      // @ts-ignore
      console.log('No accounts detected')
      return undefined
    }

    if (currentAccounts.length > 1) {
      // TBD: Add choose account code here
      // @ts-ignore
      console.log(
        'Multiple accounts detected, need to add choose account code.',
      )
      return currentAccounts[0]
    } else if (currentAccounts.length === 1) {
      return currentAccounts[0]
    }
  }
}

const ua = '' // window?.navigator?.userAgent
const msie = ua.indexOf('MSIE ')
const msie11 = ua.indexOf('Trident/')
const isIE = msie > 0 || msie11 > 0

// Log In, Log Out button
const AzureAuthenticationButton = ({ onAuthenticated }: any): JSX.Element => {
  // Azure client context
  const authenticationModule: AzureAuthenticationContext = new AzureAuthenticationContext()

  const [authenticated, setAuthenticated] = useState<Boolean>(false)
  const [user, setUser] = useState<AccountInfo>()

  const logIn = (method: string): any => {
    const typeName = 'loginPopup'
    const logInType = isIE ? 'loginRedirect' : typeName

    // Azure Login
    authenticationModule.login(logInType, returnedAccountInfo)
  }
  const logOut = (): any => {
    if (user) {
      // onAuthenticated(undefined)
      // Azure Logout
      authenticationModule.logout(user)
    }
  }

  const returnedAccountInfo = (user: AccountInfo) => {
    // set state
    setAuthenticated(user?.name ? true : false)
    // TODO: why is onAuthenticated not a function?
    // onAuthenticated(user)
    console.log(user)
    setUser(user)
  }

  const showLogInButton = (): any => {
    return (
      <button id="authenticationButton" onClick={() => logIn('loginPopup')}>
        Log in
      </button>
    )
  }

  const showLogOutButton = (): any => {
    return (
      <div id="authenticationButtonDiv">
        <div id="authentication">
          <button id="authenticationButton" onClick={() => logOut()}>
            Log out
          </button>
        </div>
      </div>
    )
  }

  const showButton = (): any => {
    return authenticated ? showLogOutButton() : showLogInButton()
  }

  return (
    <div id="authentication">
      {authenticationModule.isAuthenticationConfigured ? (
        showButton()
      ) : (
        <div>Authentication Client ID is not configured.</div>
      )}
    </div>
  )
}

export default AzureAuthenticationButton
