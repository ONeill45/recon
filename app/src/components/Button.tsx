import styled from '@emotion/styled'

export const Button = styled.button`
  padding: 8px;
  background-color: ${(props) => props.color};
  font-size: 14px;
  border-radius: 4px;
  color: black;
  font-weight: bold;
  &:hover {
    color: white;
  }
  cursor: pointer;
`
