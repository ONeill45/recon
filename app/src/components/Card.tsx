import styled from '@emotion/styled'

export const CardDiv = styled.div`
  display: flex;
  flex-direction: column;
  flex-wrap: wrap;
  justify-content: center;
  align-items: center;
  box-shadow: 0 3px 1px -2px rgb(0 0 0 / 20%), 0 2px 2px 0 rgb(0 0 0 / 14%),
    0 1px 5px 0 rgb(0 0 0 / 12%);
  height: auto;
  width: 200px;
  padding: 16px;
  margin: 16px;
`

export const CardDetailsDiv = styled.div`
  display: flex;
  flex-direction: column;
  padding: 12px;
  align-items: center;
  justify-content: center;
  text-align: center;
`
export const CardNameDiv = styled.div`
  font-size: 16px;
  font-weight: bold;
  padding: 4px;
`

export const CardDescriptionDiv = styled.div`
  font-size: 12px;
  padding: 4px;
`

export const CardDurationDiv = styled.div`
  font-size: 12px;
  padding: 4px;
  color: grey;
`

export const LogoDiv = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  border-radius: 50%;
  width: 70px;
  height: 70px;
  box-shadow: 0 3px 1px -2px rgb(0 0 0 / 20%), 0 2px 2px 0 rgb(0 0 0 / 14%),
    0 1px 5px 0 rgb(0 0 0 / 12%);
  overflow: hidden;
`
export const LogoImg = styled.img`
  height: auto;
  width: 60px;
`
