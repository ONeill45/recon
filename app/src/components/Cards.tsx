import React from 'react'
import styled from '@emotion/styled'

const CardsDiv = styled.div`
  margin: 12px;
  display: grid;
  grid-template-columns: repeat(auto-fill, minmax(300px, 1fr));
  grid-auto-rows: auto;
  grid-gap: 12px;
`

export const Cards: React.FC = ({ children }) => <CardsDiv>{children}</CardsDiv>
