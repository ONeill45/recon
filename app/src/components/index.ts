export * from './Button'
export * from './Login'
export * from './Card'
export * from './ClientCard'
export * from './Cards'
export * from './FilterCategory'
export * from './FilterPanel'
export * from './NavBar'
export * from './NavButton'
export * from './NavButtons'
export * from './ProjectCard'
export * from './NewClientForm'
export * from './PlusCircle'
export * from './SearchBar'
export * from './Spinner'
export * from './UserSelect'
