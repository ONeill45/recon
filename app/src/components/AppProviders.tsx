import React from 'react'
import '../styles/globals.css'
import { ThemeProvider } from '@emotion/react'
import { ApolloProvider } from '@apollo/client'
import { MsalProvider } from '@azure/msal-react'
import { PublicClientApplication, Configuration } from '@azure/msal-browser'
import client from '../lib/apolloClient'
import { loadEnvironmentVariable } from 'utils/functions'

const getMsalConfig = (): Configuration => ({
  auth: {
    clientId: loadEnvironmentVariable(
      process.env.NEXT_PUBLIC_AZURE_CLIENT_ID,
      'NEXT_PUBLIC_AZURE_CLIENT_ID',
    ),
    authority: `https://login.microsoftonline.com/${loadEnvironmentVariable(
      process.env.NEXT_PUBLIC_AZURE_TENANT_ID,
      'NEXT_PUBLIC_AZURE_TENANT_ID',
    )}`,
    redirectUri: 'http://localhost:3000/',
    postLogoutRedirectUri: 'http://localhost:3000/',
  },
  cache: {
    cacheLocation: 'localStorage',
  },
})
const theme = {
  fontFamily: 'Avenir Book, Arial, Verdana, sans-serif',
}

export const AppProviders: React.FC = ({ children }) => {
  const msalInstance = new PublicClientApplication(getMsalConfig())

  return (
    <MsalProvider instance={msalInstance}>
      <ThemeProvider theme={theme}>
        <ApolloProvider client={client}>{children}</ApolloProvider>
      </ThemeProvider>
    </MsalProvider>
  )
}
