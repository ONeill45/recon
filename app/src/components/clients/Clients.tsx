import React, { useEffect, useState } from 'react'
import { FaPlus } from 'react-icons/fa'
import { useLazyQuery } from '@apollo/client'

import styles from 'styles/Home.module.css'
import { Client } from 'interfaces'
import { CardsContainer } from 'components/layouts/CardsContainer'
import { ClientCard } from 'components/clients/ClientCard'
import { FilterPanel } from 'components/FilterPanel'
import { LinkButton } from 'components/common/Button'
import { PageHeader } from 'components/common/PageHeader'

import { GET_ALL_CLIENTS } from 'queries'

export const Clients: React.FC = () => {
  const [searchText, setSearchText] = useState('')
  // const [searchClients, setSearchClients] = useState<{ [key: string]: any }>([])
  const [filter, setFilter] = useState({})
  const [data, setData] = useState<{ [key: string]: any }>({})
  const [error, setError] = useState<{ [key: string]: any } | undefined>(
    undefined,
  )
  const page = 'Clients'

  const [getAllClients, { loading }] = useLazyQuery(GET_ALL_CLIENTS, {
    fetchPolicy: 'network-only',
    onCompleted: (res: { [key: string]: any }) => {
      setData(res)
    },
    onError: (err: any) => {
      setError(err)
    },
  })

  useEffect(() => {
    handleOnFilter({ searchItem: searchText }, false)
  }, [searchText])

  useEffect(() => {
    setData({})
    setError(undefined)
    getAllClients({ variables: filter })
  }, [filter, getAllClients])

  const handleOnFilter = (
    queryFilter: Record<string, any>,
    filterClicked = false,
  ) => {
    if (!queryFilter['searchItem']) {
      queryFilter['searchItem'] = searchText
    }
    setFilter((prev: any) => {
      return filterClicked ? { ...queryFilter } : { ...prev, ...queryFilter }
    })
  }

  const { clients } = data

  const renderContent = () => {
    if (loading) {
      return <p>Loading...</p>
    } else if (error) {
      return <p>Error: {error.message}</p>
    } else if (data) {
      return (
        <CardsContainer>
          {clients &&
            clients.map((client: Client) => {
              return <ClientCard key={client.id} client={client} />
            })}
        </CardsContainer>
      )
    }

    return <p></p>
  }

  return (
    <>
      <PageHeader headerText="Clients">
        <LinkButton href="/clients/client" rightIcon={<FaPlus />}>
          Create
        </LinkButton>
        <FilterPanel
          setSearchText={setSearchText}
          page={page}
          onFilter={handleOnFilter}
        />
      </PageHeader>
      <div className={styles.container}>{renderContent()}</div>
    </>
  )
}
