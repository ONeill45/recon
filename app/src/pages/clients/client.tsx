import React from 'react'
import { Client as ClientComponent } from 'components/clients/Client'

export const Client: React.FC = () => {
  return <ClientComponent />
}
export default Client
