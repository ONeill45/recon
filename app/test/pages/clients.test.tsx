import React from 'react'
import { mount } from 'enzyme'
import Clients from '../../src/pages/clients'
import axios from 'axios'
import { ClientFactory } from '../factories'
import { ClientCards } from '../../src/components'
import { act } from 'react-dom/test-utils'
import { MockedProvider } from '@apollo/client/testing'

describe('Client page test', () => {
  it('matches snapshot', () => {
    const wrapper = mount(
      <MockedProvider>
        <Clients />
      </MockedProvider>,
    )
    expect(wrapper).toMatchSnapshot()
  })

  //   it('matches snapshot', async () => {
  //     const clients = ClientFactory.buildList(5)

  //     axios.post = jest.fn().mockResolvedValue({
  //       data: {
  //         data: {
  //           clients,
  //         },
  //       },
  //     })
  //     const wrapper = mount(<Clients />)
  //     act(() => {
  //       wrapper.update()
  //     })

  //     expect(axios.post).toHaveBeenCalledWith('http://localhost:5000', {
  //       query: `
  //   {
  //     clients {
  //       id
  //       clientName
  //       description
  //       logoUrl
  //       startDate
  //       endDate
  //     }
  //   }
  // `,
  //     })

  //     await act(() => Promise.resolve())

  //     expect(wrapper.find(ClientCards).length).toEqual(1)
  //   })
})
