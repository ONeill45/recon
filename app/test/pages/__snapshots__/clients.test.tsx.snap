// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Client page test matches snapshot 1`] = `ReactWrapper {}`;
